table.cc

```diff
@@ -962,6 +962,17 @@ void table::trigger_offstrategy_compaction() {
     _compaction_manager.submit_offstrategy(this);
 }
 
+future<> table::submit_offstrategy_compaction() {
+    // If the user calls trigger_offstrategy_compaction() to trigger
+    // off-strategy explicitly, cancel the timeout based automatic trigger.
+    if (!_off_strategy_trigger.armed()) {
+        return make_ready_future<>();
```

Comment:
Is this logic correct? submit_offstrategy_compaction() should be do absolutely nothing if an offstrategy compaction wasn't already scheduled?
If this is logic is correct, I think at least the name of the function should be different. Perhaps undelay_offstrategy_compaction() or something like that.

Reply:
agreed. the function isn't really about submitting offstrategy request, but rather anticipating a offstrategy which would be latter submitted when timer expires. requires a more descriptive name. and important to mention that it's only covering offstrategy for repair. for others there's not a timer.

Reply:
That's why I liked trigger_offstrategy_compaction (if it's already scheduled, trigger it now)

Reply:
agreed. we need a name that describes it better. this function is confined to offstrategy for regular repair, and it will only submit request if there's an ongoing repair.

i tend to think this function should only perform offstrategy compaction, not check for any condition or do anything else. so it could be named run_offstrategy_compaction() instead. if it has such a generic name, it shouldn't be restricted for specific conditions or for a specific maintenance operation.

Reply:
and the logic is correct, we don't need to start a new offstrategy compaction if the trigger isn't already armed.
If it was never armed, we have nothing to compact offstrategy, If it was armed in the past and not it isn't - it means that offstrategy compaction already completed and it is not needed at the moment.

Reply:
trigger_offstrategy_compaction() is confusing if it's going to wait for execution. run_offstrategy_compaction() makes more sense. also if we check for timer, the function is confined for offstrategy on regular repair, then it shouldn't go with such a generic name. it's confusing. 

I'd suggest submit_offstrategy_compaction() to be renamed to run_offstrategy_compaction(). Then remove any condition in the function. It will only call _compaction_manager.perform_offstrategy_compaction(). If we want to make it specific to offstrategy on regular repair, then we better come up with a more descriptive name, and the API should also be documented such that its usage is confined to offstrategy on regular repair.

Reply:
We can do that.
I'll make sure it will serialize well if the timer expires in the middle of an on-going offstrategy-compaction.
(Another use case that comes into my mind is running offstrategy compaction during repair to limit the backlog of offstrategy compaction.  I'm not sure that accumulating e.g. 1024 sstables is worth it in terms of efficiency, vs. triggering offstrategy compaction periodically when e.g. 32 sstables were accumulated (per shard))

Reply:
Fixed in v4.